In [152]:
#imports
import pandas as pd

In [153]:
pd.options.display.max_columns = None

In [154]:
%config IPCompleter.greedy=True

In [155]:
#data.duplicated(subset = None, keep = "first")

In [156]:
d= "DatasetsImmo/stocks.csv"

In [157]:
data = pd.read_csv(d)

# Identification of variables and data types

In [158]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19855 entries, 0 to 19854
Data columns (total 23 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   id                             19855 non-null  int64  
 1   type                           19855 non-null  object 
 2   subtype                        19855 non-null  object 
 3   price                          19855 non-null  object 
 4   transactionType                19855 non-null  object 
 5   zip                            19855 non-null  int64  
 6   visualisationOption            19855 non-null  object 
 7   cuisine_type                   12904 non-null  object 
 8   constructionYear               11227 non-null  float64
 9   condition                      13434 non-null  object 
 10  heatingType                    12404 non-null  object 
 11  primaryEnergyConsumptionLevel  0 non-null      float64
 12  room_number                    18184 non-null 

# Finding % null values

In [159]:
data.apply(lambda x: (sum(x.isnull())/len(data))*100,axis=0)

id                                 0.000000
type                               0.000000
subtype                            0.000000
price                              0.000000
transactionType                    0.000000
zip                                0.000000
visualisationOption                0.000000
cuisine_type                      35.008814
constructionYear                  43.455049
condition                         32.339461
heatingType                       37.527071
primaryEnergyConsumptionLevel    100.000000
room_number                        8.416016
surface                           83.127676
atticExists                       86.779149
basementExists                    61.737598
hasSwimmingPool                   97.340720
indoor                            64.537900
outdoor                           72.208512
short_id                           0.000000
company_name                       5.625787
m2                                 0.000000
commune                         

# Dealing with duplicates

In [160]:
print(data.shape)
data.drop_duplicates(subset=None, keep="first",inplace=True)
print(data.shape)

(19855, 23)
(19855, 23)


Duplicates check!

In [161]:
data.head(5)

id             type          subtype            price transactionType  \
0  9042129  apartment group  apartment group  336000 - 610000        for sale   
1  9041098  apartment group  apartment group  165000 - 420000        for sale   
2  9039019        apartment        apartment          1550000        for sale   
3  9043036        apartment        apartment           195000        for sale   
4  9042073        apartment        apartment           320000        for sale   

    zip visualisationOption        cuisine_type  constructionYear condition  \
0  1000                  xl                 NaN               NaN       NaN   
1  6001                  xl                 NaN               NaN       NaN   
2  1050                  xl  usa hyper equipped            2017.0    as new   
3  9600                  xl           installed            2014.0      good   
4  1070                  xl  usa hyper equipped            2017.0    as new   

  heatingType  primaryEnergyConsumptionLevel  room_number  surface  \
0         NaN                            NaN          NaN      NaN   
1         NaN                            NaN          NaN      NaN   
2         gas                            NaN          3.0      NaN   
3         gas                            NaN          2.0      NaN   
4         gas                            NaN          3.0      NaN   

  atticExists basementExists hasSwimmingPool  indoor  outdoor  short_id  \
0         NaN            NaN             NaN     1.0      1.0   2575488   
1         NaN            NaN             NaN     NaN      NaN   1400721   
2         NaN            NaN             NaN     NaN      1.0   2595964   
3         NaN            NaN             NaN     NaN      NaN   1002634   
4         NaN           True             NaN     1.0      1.0   2916018   

                company_name   m2     commune  
0      we invest real estate   94   Bruxelles  
1       concept confort sprl   64  Marcinelle  
2  lecobel vaneau vente neuf  213     Ixelles  
3                immo beguin   75      Renaix  
4                 sky realty  130  Anderlecht

 # Exploring datas

In [162]:
data["cuisine_type"].unique()

array([nan, 'usa hyper equipped', 'installed', 'hyper equipped',
       'usa installed', 'not installed', 'semi equipped',
       'usa semi equipped', 'usa uninstalled'], dtype=object)

In [163]:
data["cuisine_type"].value_counts()

installed             6380
hyper equipped        2879
semi equipped         1333
usa hyper equipped    1155
not installed          706
usa installed          346
usa semi equipped       92
usa uninstalled         13
Name: cuisine_type, dtype: int64

In [164]:
data["constructionYear"].value_counts()

2020.0    1574
2019.0     776
2018.0     366
2021.0     342
1930.0     235
          ... 
1802.0       1
1872.0       1
1789.0       1
1786.0       1
1855.0       1
Name: constructionYear, Length: 185, dtype: int64

In [165]:
data["condition"].value_counts()

as new            5820
good              4428
to be done up     1110
to renovate       1105
just renovated     890
to restore          81
Name: condition, dtype: int64

In [166]:
data["heatingType"].value_counts()

gas         9190
fueloil     2316
electric     736
pellet       117
wood          26
carbon        14
solar          5
Name: heatingType, dtype: int64

In [167]:
data["atticExists"].value_counts()#assume rest is false

True    2625
Name: atticExists, dtype: int64

In [168]:
data["basementExists"].value_counts()#assume rest is false

True    7597
Name: basementExists, dtype: int64

In [169]:
data["hasSwimmingPool"].value_counts()#assume rest is false

True    528
Name: hasSwimmingPool, dtype: int64

In [170]:
#data["indoor"].value_counts() ??

In [171]:
#data["outdoor"].value_counts() ??

In [172]:
data.describe()

id           zip  constructionYear  \
count  1.985500e+04  19855.000000      11227.000000   
mean   8.813947e+06   5351.076958       1983.291351   
std    3.474169e+05   3203.867180         39.423541   
min    3.702839e+06   1000.000000       1755.000000   
25%    8.783413e+06   1970.000000       1962.000000   
50%    8.933765e+06   5330.000000       1993.000000   
75%    9.000254e+06   8470.000000       2018.000000   
max    9.043940e+06   9992.000000       2023.000000   

       primaryEnergyConsumptionLevel   room_number       surface  \
count                            0.0  18184.000000   3350.000000   
mean                             NaN      2.993731   1115.731940   
std                              NaN      3.426793   4881.913707   
min                              NaN      0.000000      1.000000   
25%                              NaN      2.000000     76.000000   
50%                              NaN      3.000000    210.000000   
75%                              NaN      4.000000    700.000000   
max                              NaN    204.000000  80978.000000   

             indoor      outdoor      short_id            m2  
count   7041.000000  5518.000000  1.985500e+04  19855.000000  
mean       7.673626     4.169264  1.723057e+06    152.286678  
std      417.228670    10.313800  1.207467e+06    168.879982  
min        1.000000     1.000000  1.410000e+03      1.000000  
25%        1.000000     1.000000  5.011240e+05     79.000000  
50%        1.000000     2.000000  1.838339e+06    118.000000  
75%        2.000000     4.000000  2.735512e+06    188.000000  
max    35000.000000   250.000000  3.675574e+06   8800.000000

 # Premières modifs

Removing subset of apartment group because their price is an interval instead of a definitive value

In [173]:
data.head(1)

id             type          subtype            price transactionType  \
0  9042129  apartment group  apartment group  336000 - 610000        for sale   

    zip visualisationOption cuisine_type  constructionYear condition  \
0  1000                  xl          NaN               NaN       NaN   

  heatingType  primaryEnergyConsumptionLevel  room_number  surface  \
0         NaN                            NaN          NaN      NaN   

  atticExists basementExists hasSwimmingPool  indoor  outdoor  short_id  \
0         NaN            NaN             NaN     1.0      1.0   2575488   

            company_name  m2    commune  
0  we invest real estate  94  Bruxelles

In [174]:
#data[data.subtype=="apartment group"]

---------------

 > suppressing some rows where price is an interval and  not a definitive value

In [175]:
data.type.value_counts()

house              9568
apartment          8616
apartment group    1318
house group         353
Name: type, dtype: int64

In [176]:
data = data[data.type!="apartment group"]
data = data[data.type!="house group"]
data = data[data.subtype!="apartment group"]

In [177]:
data.type.value_counts()

house        9568
apartment    8616
Name: type, dtype: int64

In [178]:
data[data.type.isin(["apartment group","house group"])]

Empty DataFrame
Columns: [id, type, subtype, price, transactionType, zip, visualisationOption, cuisine_type, constructionYear, condition, heatingType, primaryEnergyConsumptionLevel, room_number, surface, atticExists, basementExists, hasSwimmingPool, indoor, outdoor, short_id, company_name, m2, commune]
Index: []

In [179]:
data.rename(columns={'m2':"mètres carrés"},inplace=True)

deleting irrelevant columns

In [180]:
del data["surface"]
del data["primaryEnergyConsumptionLevel"]

In [181]:
data.columns

Index(['id', 'type', 'subtype', 'price', 'transactionType', 'zip',
       'visualisationOption', 'cuisine_type', 'constructionYear', 'condition',
       'heatingType', 'room_number', 'atticExists', 'basementExists',
       'hasSwimmingPool', 'indoor', 'outdoor', 'short_id', 'company_name',
       'mètres carrés', 'commune'],
      dtype='object')

### Replacing  missing value in data["atticExists"]


In [182]:
data.type.value_counts()

house        9568
apartment    8616
Name: type, dtype: int64

In [183]:
attics = data["atticExists"]

In [184]:
attics.value_counts()

True    2625
Name: atticExists, dtype: int64

- We assume that if the data is not  True it's False

In [185]:
attics.fillna(False,inplace= True)

In [186]:
attics.value_counts()

False    15559
True      2625
Name: atticExists, dtype: int64

In [187]:
data["atticExists"] =attics

In [188]:
data["atticExists"].value_counts()

False    15559
True      2625
Name: atticExists, dtype: int64

#### Doing the same with data["hasSwimmingPool"] and data["basementExists"]
#### and data["cuisine_type"]

In [189]:
data["hasSwimmingPool"].fillna(False,inplace= True)
data["basementExists"].fillna(False,inplace= True)

## Let's see :

In [190]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 18184 entries, 2 to 19854
Data columns (total 21 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   id                   18184 non-null  int64  
 1   type                 18184 non-null  object 
 2   subtype              18184 non-null  object 
 3   price                18184 non-null  object 
 4   transactionType      18184 non-null  object 
 5   zip                  18184 non-null  int64  
 6   visualisationOption  18184 non-null  object 
 7   cuisine_type         12904 non-null  object 
 8   constructionYear     11227 non-null  float64
 9   condition            13434 non-null  object 
 10  heatingType          12404 non-null  object 
 11  room_number          18184 non-null  float64
 12  atticExists          18184 non-null  bool   
 13  basementExists       18184 non-null  bool   
 14  hasSwimmingPool      18184 non-null  bool   
 15  indoor               6783 non-null  

#### condition, heatingtype,room_number and company_name need to be dealt with

# The dropna approach

### Removing rows where  heatingType , condition and room_number are missing

In [191]:
data_with_removeNAN = data.copy()

In [192]:
data_with_removeNAN.shape

(18184, 21)

In [193]:
data_with_removeNAN.dropna(subset=["heatingType","condition","room_number","cuisine_type","constructionYear"],inplace=True)

In [194]:
data_with_removeNAN.shape

(5387, 21)

# Transforming date to int

In [195]:
data_with_removeNAN["constructionYear"].astype("int64")

2        2017
3        2014
4        2017
5        2020
6        2013
         ... 
19626    2019
19636    2020
19669    2020
19705    1964
19736    2020
Name: constructionYear, Length: 5387, dtype: int64

In [196]:
#data_with_removeNAN["indoor"].value_counts()

 - ^After looking up indoor and outdoor are values corresponding to parking spots. So lets rename them for clarity sake

In [197]:
data_with_removeNAN.rename(columns={'outdoor':"parking extérieur"},inplace=True)

In [198]:
data_with_removeNAN.rename(columns={'indoor':"parking intérieur"},inplace=True)

- Still need  to replace missing values in those two.

In [199]:
data_with_removeNAN["parking extérieur"].fillna(0,inplace=True)
data_with_removeNAN["parking intérieur"].fillna(0,inplace=True)

## Checking our data again

In [200]:
data_with_removeNAN.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5387 entries, 2 to 19736
Data columns (total 21 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   id                   5387 non-null   int64  
 1   type                 5387 non-null   object 
 2   subtype              5387 non-null   object 
 3   price                5387 non-null   object 
 4   transactionType      5387 non-null   object 
 5   zip                  5387 non-null   int64  
 6   visualisationOption  5387 non-null   object 
 7   cuisine_type         5387 non-null   object 
 8   constructionYear     5387 non-null   float64
 9   condition            5387 non-null   object 
 10  heatingType          5387 non-null   object 
 11  room_number          5387 non-null   float64
 12  atticExists          5387 non-null   bool   
 13  basementExists       5387 non-null   bool   
 14  hasSwimmingPool      5387 non-null   bool   
 15  parking intérieur    5387 non-null   

## What to do ?

A few hundreds missing values still exist in "company name". We could assume that those missing values are because the sellers are not agency but individuals or we could just simply delete the rows with missing datas. SInce we opted for removing the rows with missing values in the other columns, lets stay consistent.

In [201]:
data_with_removeNAN.dropna(subset=["company_name"],inplace=True)

In [202]:
data_with_removeNAN.shape

(4918, 21)

In [203]:
data_with_removeNAN.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4918 entries, 2 to 19736
Data columns (total 21 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   id                   4918 non-null   int64  
 1   type                 4918 non-null   object 
 2   subtype              4918 non-null   object 
 3   price                4918 non-null   object 
 4   transactionType      4918 non-null   object 
 5   zip                  4918 non-null   int64  
 6   visualisationOption  4918 non-null   object 
 7   cuisine_type         4918 non-null   object 
 8   constructionYear     4918 non-null   float64
 9   condition            4918 non-null   object 
 10  heatingType          4918 non-null   object 
 11  room_number          4918 non-null   float64
 12  atticExists          4918 non-null   bool   
 13  basementExists       4918 non-null   bool   
 14  hasSwimmingPool      4918 non-null   bool   
 15  parking intérieur    4918 non-null   

- We finally arrive with a dataset with 4918 lines where all values are filled !

# The filling approach

In [204]:
data_with_fillNa = data.copy()

In [205]:
data_with_fillNa["heatingType"].fillna(method='pad',inplace=True)
data_with_fillNa["condition"].fillna(method='pad',inplace=True)
data_with_fillNa["room_number"].fillna(method='pad',inplace=True)
data_with_fillNa["cuisine_type"].fillna("No infos",inplace= True)

In [206]:
data_with_fillNa["company_name"].fillna("Individuals",inplace=True)

In [207]:
construction = data["constructionYear"]
construction.fillna(data_with_fillNa["constructionYear"].mean(),inplace=True)
data_with_fillNa["constructionYear"] = construction

Quick redo of indoor/outdoor stuff

In [208]:
data_with_fillNa.rename(columns={'outdoor':"parking extérieur"},inplace=True)
data_with_fillNa.rename(columns={'indoor':"parking intérieur"},inplace=True)
data_with_fillNa["parking extérieur"].fillna(0,inplace=True)
data_with_fillNa["parking intérieur"].fillna(0,inplace=True)

In [209]:
data_with_fillNa.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 18184 entries, 2 to 19854
Data columns (total 21 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   id                   18184 non-null  int64  
 1   type                 18184 non-null  object 
 2   subtype              18184 non-null  object 
 3   price                18184 non-null  object 
 4   transactionType      18184 non-null  object 
 5   zip                  18184 non-null  int64  
 6   visualisationOption  18184 non-null  object 
 7   cuisine_type         18184 non-null  object 
 8   constructionYear     18184 non-null  float64
 9   condition            18184 non-null  object 
 10  heatingType          18184 non-null  object 
 11  room_number          18184 non-null  float64
 12  atticExists          18184 non-null  bool   
 13  basementExists       18184 non-null  bool   
 14  hasSwimmingPool      18184 non-null  bool   
 15  parking intérieur    18184 non-null 

drop those 2 lines with some missing values :

In [210]:
data_with_fillNa.dropna(inplace=True)

# Export those 2 in too differents csv for further analysis

In [211]:
data_with_removeNAN.to_csv("DatasetsImmo/clean_remove_approach.csv")

In [212]:
data_with_fillNa.to_csv("DatasetsImmo/clean_fill_approach.csv")